### PR TITLE
Proposed handling of Shopify API versioning

### PR DESCRIPTION
**Pull request by dougyss4**
One potential path toward utilizing the api version in Shopify admin URLs.

### CODE REVIEW

File: goshopify.go [hunk 1]
```diff
@@ -424,3 +431,14 @@ func (c *Client) Put(path string, data, resource interface{}) error {
 func (c *Client) Delete(path string) error {
 	return c.CreateAndDo("DELETE", path, nil, nil, nil)
 }
+
+func GetAdminVersionedApiPathPrefix() string {
```
Comment (rickywiens) [inline goshopify.go]: Is there a reason we have to re-apply the validation on every call to this function? I think if we could validate the apiVersion a single time, then just return either admin, or the compiled version string that would be more performant.

File: goshopify_test.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ func setup() {
 		Scope:       "read_products",
 		Password:    "privateapppassword",
 	}
-	client = NewClient(app, "fooshop", "abcd")
+	client = NewClient(app, "fooshop", "abcd", "2019-04")
```
Comment (rickywiens) [inline goshopify_test.go]: I wonder if we should hard-code the API version in the tests to be an invalid y/d but pass the validation so it's clearly a test value? 0000-00 or 9999-99 or something?

File: applicationcharge.go [hunk 3]
```diff
@@ -39,41 +39,41 @@ type ApplicationCharge struct {
 }
 
 // ApplicationChargeResource represents the result from the
-// admin/application_charges{/X{/activate.json}.json}.json endpoints.
+// admin/api/2019-04/application_charges{/X{/activate.json}.json}.json endpoints.
```
Comment (rickywiens) [inline applicationcharge.go]: Was this change intentional? Same with the one on line 48?

File: applicationcharge_test.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ func applicationChargeTests(t *testing.T, charge ApplicationCharge) {
 		},
 		{
 			"ConfirmationURL",
-			"https://apple.myshopify.com/admin/charges/1017262355/confirm_application_charge?sign" +
+			"https://apple.myshopify.com/admin/api/2019-04/charges/1017262355/confirm_application_charge?sign" +
```
Comment (ekhabarov) [inline applicationcharge_test.go]: Next ShopifyAPI version will be `2019-07`. I think here `GetAdminVersionedApiPathPrefix()` function could be used for reducing size of next PR and decreasing number of changes.

Comment (rickywiens) [reply applicationcharge_test.go]: If you see my comment here:

https://github.com/bold-commerce/go-shopify/pull/48#discussion_r277762885

Which one would you prefer?

File: goshopify.go [hunk 5]
```diff
@@ -424,3 +431,14 @@ func (c *Client) Put(path string, data, resource interface{}) error {
...
+	// 	if our apiVersion exists, this should return the form:  "admin/api/2019-04"
+	// 	if not, this should return the form:  "admin/", as the default year ago version going forward.
+	var rxPat = regexp.MustCompile(`^[0-9]{4}-[0-9]{2}$`)
+	if len(apiVersion) > 0 && rxPat.MatchString(apiVersion) {
+		return fmt.Sprintf("admin/api/%s", apiVersion)
+	} else {
+		return "admin"
```
Comment (rickywiens) [inline goshopify.go]: This line will also need to get hit with a test (that's why the coverage dropped).

File: goshopify.go (hunk 1 above)
Comment (dougyss4) [reply goshopify.go]: Cleanest path I see for this is to migrate that functionality directly into the NewClient function, and then access the data via a variable everywhere else for accessing the path prefix.

File: applicationcharge.go (hunk 3 above)
Comment (dougyss4) [reply applicationcharge.go]: I had made it while passing over everything and trying to address the overall version issue, not really sure if it's more or less clear in a comment like this over time...  I've reverted both lines that you've brought up here.

File: goshopify_test.go (hunk 2 above)
Comment (dougyss4) [reply goshopify_test.go]: Given the pattern change I'm ready to commit the test value would be provided in the NewClient call.  There are a few cases where the value is present in the fixtures, I've set everything at 9999-99

File: goshopify.go (hunk 5 above)
Comment (dougyss4) [reply goshopify.go]: Kind of skirted this by getting rid of the function, probably should expand the NewClient test to include invalid version input testing.

File: goshopify.go [hunk 6]
```diff
@@ -177,14 +183,24 @@ func (c *Client) NewRequest(method, urlStr string, body, options interface{}) (*
 // token. The shopName parameter is the shop's myshopify domain,
```
[the remark this answers is not in the document]
Comment (gordcurrie) [reply goshopify.go]: usage example
```
	client := goshopify.NewClient(app, "my-shop", "not-a-real-token", goshopify.WithVersion("2019-19"))
```

[the remark this answers is not in the document]
Comment (rickywiens) [reply goshopify.go]: I like this too.

[the remark this answers is not in the document]
Comment (dougyss4) [reply goshopify.go]: Appreciated, this change has now  been pushed.

[the remark this answers is not in the document]
Comment (gordcurrie) [reply goshopify.go]: Thank you for the work you've put in, looks good.
